### PR TITLE
Image georeferencer

### PR DESCRIPTION
**Pull request by caballeto**
## Prototype of image georeferencer
In order to create an accurate image projection (or map overlay) we have to map the image to certain Coordinate Reference System. This will allow us to reference the Internal CS with the CS of the projection. This transformation requires two important parts:
- Accurate computation of GCPs (Ground Control Points)
- Efficient software for georeferencing image with GCPs

The 2 part can be done using the API of `gdal` library. It provides all needed functionality to properly georeference the image.

The 1 task is far more challengeable. The coordinates of the computed GCPs should be very accurate, because even slight distortion will result in badly georeferenced image. This implementation computes all GCPs automatically, but the computation algorithm is rather simple and doesn't yet provide needed accuracy. Current result is presented below.

Bad

The result image is not accurate because of the several assumptions that algorithm makes:
- every pixel has the same number of kilometers
- don't take in account the distortion near the image border
- assume that all the angles on the projection are equal to the angles on the picture

These assumptions plus an error which adds up while calculating distances between points creates such big mismatches with the real projection. To deal with the problems above every parameter should be finetuned and a certain accurate dependency should be calculated between APT image and the projection. This mapping will depend on APT format.

Should be noted, that the generated image can't be visualized in `ImageMerger` because cartopy does not support `EPSG:4326`, QGIS program should be used instead.

To sum up, this implementation presents a working routine for image georeferencing, with inaccurate computation of GCP points. Below is presented an image with few, yet accurate GCP points (chosen manually).

overlay

I want to emphasize, that solving the problem of accurately assigning geocoordinates to some pixels (computing GCPs) is the first step for creating a good-looking globe projection, which will require assigning accurate coordinates to every pixel on the image.

**Edit.** The final computation algorithm is as follows: 

- load tle file 
- load descriptor file 
- extract image, date of observation 
- compute gcps 
- translate image 
- warp image 
- create descriptor if `desc` flag is set 

We compute ground control points in several steps. First we extract the datetime of observation and calculate the position of satellite 5 seconds before the start, it is stored in prev variable. This is done, because we will use it to compute the azimuth angle of for each line we iterate over. The next step is the loop, where on each iteration we compute 5 ground control points, namely point on the left boundary, point on quarter, point in the middle, point ¾ quarter, and point on right boundary. To get these, we: 

- calculate the gcp_time – time when satellite will be at this point 
- calculate the position of the satellite at this point 
- compute the azimuth angle between middle point, and the perpendicular to satellite orbit at this point 
- using GeographicLib we compute 4 other points, knowing the azimuth angle (direction) and the distance we calculate the needed position

We then pass these points to gdal routines and receive the georeferenced image.  
Below is an example of the result image.  

Screenshot_20190410_233050

### CODE REVIEW

File: directdemod/constants.py [hunk 1]
```diff
@@ -25,6 +25,8 @@
 ## Merger settings
 RESOLUTION = 500
 COLOR = "black"
+TEMP_TIFF_FILE = "_temp.tiff"
```
Comment (7andahalf) [inline directdemod/constants.py:28]: Just curious, why .tiff? Are you just storing image or there is some metadata also?

File: directdemod/constants.py [hunk 2]
```diff
@@ -25,6 +25,8 @@
...
+DEFAULT_RS = "EPSG:4326"
```
Comment (7andahalf) [inline directdemod/constants.py:29]: Is this compatible with WGS84? Is this the same reference frame as what pyorbital outputs?

File: directdemod/georeferencer.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        height = image.shape[0]
+        width  = image.shape[1]
+        center_w = width/2
+        center_h = height/2
+
+        gcps = []
+        dtime = dparser.parse(descriptor["date_time"])
+        orbiter = Orbital(descriptor["sat_type"], tle_file=self.tle_file)
+        min_delta = 500
+        middle_dist = 3.22 * 455 / 2. * 1000
```
Comment (7andahalf) [inline directdemod/georeferencer.py:157]: could you please explain what these middle_dist and far_dist are? how did you arrive at these numbers?

File: directdemod/json_parser.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
+'''
+Object for json manipulation
```
Comment (7andahalf) [inline directdemod/json_parser.py]: If possible put it in misc.py, I just want the directDemod folder to be mostly "comm" stuff.

File: directdemod/merger.py [hunk 5]
```diff
@@ -92,19 +92,23 @@ def merge(self, jsons, whole=False, resolution=constants.RESOLUTION):
...
         for obj in jsons:
-            image  = mimg.imread(obj["image_name"])
+            image, isGray = self.imread(obj["image_name"])
             center = obj["center"]
             degree = obj["direction"]
 
             if self.is_cartopy:
-                img = self.set_transparent(ndimage.rotate(image, degree, cval=255))
+                img = self.set_transparent(ndimage.rotate(image, degree, cval=255), isGray)
                 dx = img.shape[0]*4000/2*0.81 # in meters
```
Comment (7andahalf) [inline directdemod/merger.py:101]: This arbitrary 0.81 still remains even after georeferencing? or this is something else?

File: directdemod/georeferencer.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    def create_desc(self, descriptor, output_file):
+
+        '''Create descriptor file
+
+        Args:
+            descriptor (:obj:`dict`): descriptor dictionary
+            output_file (:obj:`string`): name of the output file
+        '''
+
+        desc = {
```
Comment (7andahalf) [inline directdemod/georeferencer.py:127]: hmmm.. this descriptor is still needed?
We could use an image format which will store all this info within itself. Did you take a look at geotiff? is it possible to store the needed info within the image. Actually, the only info that the image needs is: each pixel should be associated with a lat/long and none of these other things are required. We have to change the NOAA decoder part which produces the image such that when it writes the image it also create this image which has every pixel tagged with a latlong. By doing this, we can make our code independent of which sat is producing the image. If I want to use a meteor image, that can also produce an image with tagged pixels and I can just use that image also. I suspect further that even there will be variation within the 3 different NOAA sats.

**Review by 7andahalf — COMMENTED** [review]
Hey, Thanks! This is great progress :D

I just had a few points regarding the code organization, I just breezed through the code, so I might be wrong, please correct me. Also, ignore these for now and focus on making the projection better. We will deal with all this "formality" later:
1. I feel the georeferencing has to be done within the image decoder, say NOAA decoder decodes the image and writes it in a format where all the pixels are already tagged with their lat/longs. So that we can make every decoder independent but still be able to use same merger module. So, in the end we should use an image format which allows this and no descriptor file is needed.
2. I think the rest follows naturally, if you have an image with tagged pixels and you want to show it, you can use GDAL to warp it using those pixels as GCPs as needed and display it. 
3. Merger also becomes simple as you can now do operations like spatial average etc.

What do you think? Is this how the code is already structured?

File: directdemod/merger.py (hunk 5 above)
Comment (caballeto) [reply directdemod/merger.py:101]: I have written in log file, that merger isn't yet made to support plotting of tiff files. Because I thought about using something more powerful. So in this PR merger hasn't been changed, I only added support of gray images.

File: directdemod/json_parser.py (hunk 4 above)
Comment (caballeto) [reply directdemod/json_parser.py]: Okey.

File: directdemod/constants.py (hunk 1 above)
Comment (caballeto) [reply directdemod/constants.py:28]: This is for gdal routines, when gdal translate is called it by default creates a temp copy png and xml file with gpc points. I changed this behaviour, now when gdal translate is called it creates a tiff file where it stores info about coordinate system, gcp points, image itself and other metadata. Then when gdal warp is called it transforms this tiff so the gcps match the projection. Warp creates needed output file, and temp file is deleted.
So to sum up .tiff is used in order to store all info in one place and work only with one file at the same time.

File: directdemod/constants.py (hunk 2 above)
Comment (caballeto) [reply directdemod/constants.py:29]: EPSG:4326 is just a synonym of WGS84, these names are equal

File: directdemod/georeferencer.py (hunk 3 above)
Comment (caballeto) [reply directdemod/georeferencer.py:157]: middle_dist is quarter length of image. So image an image with height 1000 and width 910, so the quarter is 455 / 2 pixels, and I multiply it by 3.22 * 1000, 3.22 km in pixel and convert it to meters. far_dist is the half of the image.
If you stay in the middle of the image, far_dist is the distance to the left and right sides, middle_dist is half of that distance.

File: directdemod/georeferencer.py (hunk 6 above)
Comment (caballeto) [reply directdemod/georeferencer.py:127]: Here as you may have notice the generation of descriptor is optional, and I think not needed. 
After georeferencing actually all info is stored in the tiff file. So descriptor is only needed before georeferencing. Just added for the case, but it may be deleted.
I think that if decoders produce descriptor files, it would be more flexible, because maybe not all users would like to georeference the image etc; in addition descriptor can save all the info about the image, including date (tiff doesn't store it), satellite name and so on. So here we just create descriptors and then people could do whatever they want, without any limitations. And there won't be differences between the decoders results (i. e. they will be independent, I described that in the proposal).
